src/freenet/crypt/SHA256.java

```diff
@@ -101,9 +87,9 @@ public static MessageDigest getMessageDigest() {
 		try {
 			MessageDigest md = null;
 			synchronized(digests) {
-				int x = digests.size();
-				if(x == 0) md = null;
-				else md = digests.remove(x-1);
+				while (digests.size() > 0 && md == null) {
+					md = digests.remove(0).get();
+				}
```

Comment:
From an algorithmic perspective I was thinking of `ConcurrentLinkedQueue`, but you are right in that it wastes some memory with an upside of possibly reducing thread contention. But just a simple ArrayDeque with synchronization will probably do as well, I don't think locking is a true bottle-neck here.